controllers/internal/utils/dns.go

```diff
@@ -20,30 +20,99 @@ package utils
 import (
 	"fmt"
 	"sort"
+	"strings"
 
 	"github.com/lixiangzhong/dnsutil"
+	"github.com/miekg/dns"
 )
 
+type DNSServer struct {
+	Host string
+	Port int
+}
+
 // Dig retrieves list of tuple <IP address, A record > from edge DNS server for specific FQDN
 func Dig(edgeDNSServer, fqdn string) ([]string, error) {
-	var dig dnsutil.Dig
 	if edgeDNSServer == "" {
 		return nil, fmt.Errorf("empty edgeDNSServer")
 	}
-	err := dig.SetDNS(edgeDNSServer)
-	if err != nil {
-		err = fmt.Errorf("dig error: can't set query dns (%s) with error(%s)", edgeDNSServer, err)
-		return nil, err
+	dnsServer := DNSServer{
+		Host: edgeDNSServer,
+		Port: 53,
+	}
+	return DigWithFallback([]DNSServer{dnsServer}, fqdn)
+}
+
+func DigWithFallback(edgeDNSServers []DNSServer, fqdn string) ([]string, error) {
```

Comment:
Please, could you describe, what should be result of function and how it works?

Basically it's very hard to read implementation here.  Is it necessary to distinguish between odd and even elements? 

Many states can occur in this function, should be tested for  possible corner cases (currently tested for not nil and result[0]) .

Reply:
good catch, I will add more tests and some comment

Reply:
actually it's being tested for 3 elements and 4 elements, in dns_test.go (`TestOneValidEdgeDNSInTheList` and `TestNoValidEdgeDNSInTheList`)